massa-grpc/src/api.rs

```diff
@@ -0,0 +1,894 @@
+                            }
+
+                            (proto_nt_block_draws, proto_nt_endorsement_draws)
+                        })
+                        .collect::<Vec<_>>()
+                };
+
+                // compile results
+                let mut res = Vec::with_capacity(addresses.len());
+                let iterator = izip!(addresses.into_iter(), selection_draws.into_iter());
+                for (address, (next_block_draws, next_endorsement_draws)) in iterator {
+                    res.push(grpc::SelectorDraws {
+                        address: address.to_string(),
+                        next_block_draws,
+                        next_endorsement_draws,
+                    });
+                }
+
+                Ok(tonic::Response::new(grpc::GetSelectorDrawsResponse {
+                    id,
+                    selector_draws: res,
+                }))
+            }
+
+            Err(e) => {
+                return Err(tonic::Status::invalid_argument(e.to_string()));
+            }
+        }
+    }
+
+    /// Get multiple datastore entries.
+    async fn get_datastore_entries(
+        &self,
+        request: tonic::Request<grpc::GetDatastoreEntriesRequest>,
+    ) -> Result<tonic::Response<grpc::GetDatastoreEntriesResponse>, tonic::Status> {
+        let execution_controller = self.execution_controller.clone();
+        let inner_req = request.into_inner();
+        let id = inner_req.id.clone();
+
+        let filters = inner_req
+            .entries
+            .into_iter()
+            .map(|filter| {
+                Address::from_str(filter.address.as_str()).map(|address| (address, filter.key))
+            })
+            .collect::<Result<Vec<_>, _>>();
+
+        match filters {
+            Ok(filters) => {
+                let entries = execution_controller
+                    .get_final_and_active_data_entry(filters)
+                    .into_iter()
+                    .map(|output| grpc::BytesMapFieldEntry {
+                        //TODO this behaviour should be confirmed
+                        key: output.0.unwrap_or_default(),
+                        value: output.1.unwrap_or_default(),
+                    })
+                    .collect();
+
+                Ok(tonic::Response::new(grpc::GetDatastoreEntriesResponse {
+                    id,
+                    entries,
+                }))
+            }
+
+            Err(e) => {
+                return Err(tonic::Status::invalid_argument(e.to_string()));
+            }
+        }
+    }
+
+    async fn get_version(
+        &self,
+        request: tonic::Request<grpc::GetVersionRequest>,
+    ) -> Result<tonic::Response<grpc::GetVersionResponse>, tonic::Status> {
+        Ok(tonic::Response::new(grpc::GetVersionResponse {
+            id: request.into_inner().id,
+            version: self.version.to_string(),
+        }))
+    }
+
+    type SendBlocksStream = Pin<
```

Comment:
put type declaration on top

Reply:
Yep, It will be more readable. What do you think about splitting this file into different services ? 

Block service, Operations service, Endorsements service, Node service... etc

Reply:
yes we can do it like that, or in first time split stream and no stream in 2 file. 
Or as you want

Reply:
We couldn't do this. It's a part from the generated Trait. they should be implemented inside